[4.1] Revert removal of some extensions on ImplicitlyUnwrappedOptional<T>

rudkx opened the pull request:
It breaks bridging to Objective C.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

slavapestov commented:
Why does it break bridging?

rudkx commented:
@slavapestov Because the change removed the extension to `_ObjectiveCBridgeable`.

rudkx commented:
Linux failure is unrelated to my changes:
```
clang-5.0: error: unable to execute command: Bus error (core dumped)
clang-5.0: error: linker command failed due to signal (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 254 (use -v to see invocation)
```

rudkx commented:
@swift-ci Please smoke test Linux platform

rudkx commented:
The SIL parsing failure looks like a result of https://github.com/apple/swift/pull/12430.

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

DougGregor commented on stdlib/public/core/ImplicitlyUnwrappedOptional.swift:
```diff
@@ -49,3 +49,73 @@ public enum ImplicitlyUnwrappedOptional<Wrapped> : ExpressibleByNilLiteral {
   }
 }
+
+extension ImplicitlyUnwrappedOptional : CustomStringConvertible {
```
Should we leave this one "removed"? It's not needed to fix the regression at hand.

DougGregor commented on stdlib/public/core/ImplicitlyUnwrappedOptional.swift:
```diff
@@ -49,3 +49,73 @@ public enum ImplicitlyUnwrappedOptional<Wrapped> : ExpressibleByNilLiteral {
+/// _isOptional thus cannot distinguish ImplicitlyUnwrappedOptional
+/// from Optional. When conditional conformance is available, this
+/// outright conformance can be removed.
+extension ImplicitlyUnwrappedOptional : CustomDebugStringConvertible {
```
Should we leave this one "removed"? It's not needed to fix the regression at hand.

DougGregor commented on validation-test/SIL/parse_stdlib.sil:
```diff
@@ -3,3 +3,12 @@
+// methods, we're fine, but in https://github.com/apple/swift/pull/13976
+// I am reintroducing extensions because removing them broke things
+// (bridging to ObjectiveC, for one).
+// REQUIRES: disabled
```
If we remove the two extensions I flagged above, we could leave this test enabled for Linux.

rudkx replied:
Yes, true. I'll look at that. I was just hoping to revert and move forward, but if we're confident we won't hit other issues from these others being removed then we should do that.

Commit:
Restore ImplicitlyUnwrappedOptional extension conforming to _ObjectiveCBridgeable.

Partial revert of 26f6a751c4dc97383f35bc2d913317df15b54c15.

Removing this breaks bridging these values to Objective C. Once IUOs
are removed from the type system and ImplicitlyUnwrappedOptional<T> is
removed from the library, we'll be strictly using Optional<T> which
has this conformance as well.

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
The source compat failure was a result of a problem cloning from GitHub:
```
error: failed to clone; Cloning into bare repository '/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/R.swift/.build/repositories/Commander.git-8842944228949165507'...
fatal: unable to access 'https://github.com/kylef/Commander.git/': Could not resolve host: github.com
```

rudkx commented:
@swift-ci Please test source compatibility
